### ISSUE TITLE
‼️ Notice: CDK CLI: "The security token included in the request is invalid" when using user credentials since 2.167.0

### ISSUE THREAD

**Issue by AdamPD** [aws/aws-cdk#32120]
Please add your +1 👍 to let us know you have encountered this
---

### Overview:

In version `2.167.0` CLI commands, including deployments, fail with authentication errors. 

This main cause is with authentication configurations that do not use `AWS_SESSION_TOKEN`, like IAM User credentials.
A second cause is related to the location of the region configuration, see https://github.com/aws/aws-cdk/issues/32130

#### Complete Error Message:

```
The security token included in the request is invalid
```

### Workaround:

Revert to `2.166.0`

### Solution:

n/a

### Related Issues:

https://github.com/aws/aws-cdk/issues/32130

---

### Original issue

Since version 2.167.0, deployments fail due to the inability to get the AWS account ID.  The following error occurs in the verbose output of cdk synthesize/deploy:

```
[01:50:17] Resolving default credentials
[01:50:17] Looking up default account ID from STS
[01:50:18] Unable to determine the default AWS account (InvalidClientTokenId): The security token included in the request is invalid
```

This is despite having correct AWS environment variables set (AWS_DEFAULT_REGION, AWS_SECRET_ACCESS_KEY, AWS_ACCESS_KEY_ID) and `aws sts get-caller-identity` works correctly.

Rolling back to 2.166.0 with:

```
npm install -g cdk@2.166.0
```

resolves the issue and deployments resume as per normal.

### Regression Issue

- [x] Select this option if this issue appears to be a regression.

### Expected Behavior

CDK should function correctly and retrieve the account ID via the AWS credentials.

### Current Behavior

CDK throws an error stating that it cannot retrieve the account ID due to security token issues.

### Reproduction Steps

Upgrade to 2.167.0, use AWS environment variable credentials but don't specify the account ID, and run `cdk synthesize`.

### CDK CLI Version

2.167.0 (build 677e108)

### Node.js Version

v22.6.0

### OS

Ubuntu 22.04.4 LTS

### Language

.NET

**Comment by Eonfuzz** [comment]
It's affecting our Azure cdk deploy pipelines too.

**Comment by christiandunn** [comment]
I have noticed this issue on our CI/CD pipelines as well. AWS credential configuration that has worked before has stopped working.

**Comment by IgorPietraszko** [comment]
Same here.  Downgraded to 2.166.0 and trying out.

**Comment by minimum-hsu** [comment]
In CDK 2.167.0, got an error meesage

```
botocore.exceptions.ClientError: An error occurred (UnrecognizedClientException) when calling the DescribeRepositories operation: The security token included in the request is invalid
```

After downgrading, the cdk diff and deploy jobs are functioning correctly.

**Comment by christiandunn** [comment]
Downgrading to the previous version also fixed the problem for me.

**Comment by raducostea** [comment]
Same for me, i have aws environment set up correctly. Downgrading to 2.166.0 fixed the issue.
```
Need to perform AWS calls for account ***, but no credentials have been configured
```

**Comment by otaviomacedo** [comment]
@AdamPD (or anyone else in this thread) can you please paste the output of `cdk deploy -vvv`?

**Comment by scarytom** [comment]
@otaviomacedo here is the -vvv output from a cdk diff, which works fine on 2.166.0
```shell
[09:08:41] [trace] SdkProvider#resolveEnvironment()
[09:08:41] [trace] SdkProvider#baseCredentialsPartition()
[09:08:41] [trace]   SdkProvider#resolveEnvironment()
[09:08:41] [trace]   SdkProvider#obtainBaseCredentials()
[09:08:41] [trace]     SdkProvider#defaultAccount()
[09:08:41] [trace]     SdkProvider#defaultCredentials()
[09:08:41] [trace]   SDK#currentAccount()
[09:08:41] Retrieved account ID YYYYYYYYYYYY from disk cache
[09:08:41] [trace] SdkProvider#forEnvironment()
[09:08:41] [trace]   SdkProvider#resolveEnvironment()
[09:08:41] [trace]   SdkProvider#obtainBaseCredentials()
[09:08:41] [trace]     SdkProvider#defaultAccount()
[09:08:41] [trace]     SdkProvider#defaultCredentials()
[09:08:41] [trace]   SdkProvider#withAssumedRole()
[09:08:41] Assuming role 'arn:aws:iam::XXXXXXXXXXXX:role/cdk-sbsl-infra-lookup-role-XXXXXXXXXXXX-us-east-1'.
[09:08:42] Assuming role failed: The security token included in the request is invalid
Could not assume role in target account using current credentials (which are for account YYYYYYYYYYYY) The security token included in the request is invalid . Please make sure that this role exists in the account. If it doesn't exist, (re)-bootstrap the environment with the right '--trust', using the latest version of the CDK CLI.
[09:08:42] [trace] SdkProvider#resolveEnvironment()
[09:08:42] [trace] SdkProvider#baseCredentialsPartition()
[09:08:42] [trace]   SdkProvider#resolveEnvironment()
[09:08:42] [trace]   SdkProvider#obtainBaseCredentials()
[09:08:42] [trace]     SdkProvider#defaultAccount()
[09:08:42] [trace]     SdkProvider#defaultCredentials()
[09:08:42] [trace]   SDK#currentAccount()
[09:08:42] Retrieved account ID YYYYYYYYYYYY from disk cache
[09:08:42] [trace] SdkProvider#forEnvironment()
[09:08:42] [trace]   SdkProvider#resolveEnvironment()
[09:08:42] [trace]   SdkProvider#obtainBaseCredentials()
[09:08:42] [trace]     SdkProvider#defaultAccount()
[09:08:42] [trace]     SdkProvider#defaultCredentials()
[09:08:42] [trace]   SdkProvider#withAssumedRole()
[09:08:42] Assuming role 'arn:aws:iam::XXXXXXXXXXXX:role/cdk-sbsl-infra-deploy-role-XXXXXXXXXXXX-us-east-1'.
[09:08:43] Assuming role failed: The security token included in the request is invalid
[09:08:43] Reading cached notices from /home/tom/.cdk/cache/notices.json
Could not assume role in target account using current credentials (which are for account YYYYYYYYYYYY) The security token included in the request is invalid . Please make sure that this role exists in the account. If it doesn't exist, (re)-bootstrap the environment with the right '--trust', using the latest version of the CDK CLI.
[09:08:43] Error: Could not assume role in target account using current credentials (which are for account YYYYYYYYYYYY) The security token included in the request is invalid . Please make sure that this role exists in the account. If it doesn't exist, (re)-bootstrap the environment with the right '--trust', using the latest version of the CDK CLI.
    at SdkProvider.withAssumedRole (/usr/lib/node_modules/aws-cdk/lib/index.js:593:5643098)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async SdkProvider.forEnvironment (/usr/lib/node_modules/aws-cdk/lib/index.js:593:5639480)
    at async EnvironmentAccess.cachedSdkForEnvironment (/usr/lib/node_modules/aws-cdk/lib/index.js:647:7329)
    at async EnvironmentAccess.prepareSdk (/usr/lib/node_modules/aws-cdk/lib/index.js:647:6285)
    at async Deployments.readCurrentTemplateWithNestedStacks (/usr/lib/node_modules/aws-cdk/lib/index.js:647:8697)
    at async CdkToolkit.diff (/usr/lib/node_modules/aws-cdk/lib/index.js:647:201596)
    at async exec3 (/usr/lib/node_modules/aws-cdk/lib/index.js:650:18965)
```

**Comment by otbe** [comment]
For us the issues is in the region, e.g.

deploy -vvv
```
10:38:42] Looking up AWS region in the EC2 Instance Metadata Service (IMDS).
[10:38:43] Unable to retrieve AWS region from IMDS: Error: Error fetching metadata token: TimeoutError: Connection timed out after 1000 ms
[10:38:43] Unable to determine AWS region from environment or AWS configuration (profile: "default"), defaulting to 'us-east-1'
```

Not quite sure how it worked in older aws-cdk versions but the region is specified in `.aws/credentials` similar to this

```
[default]
aws_access_key_id=
aws_secret_access_key=
aws_session_token=
aws_security_token=
region=eu-central-1
EOF
```

**Comment by rix0rrr** [comment]
Do you also have a `~/.aws/config` file?

**Comment by AdamPD** [comment]
To be clear on my side, I am running this within AppVeyor on a clean build each time, and we do not have any `~/.aws` directory with any configuration - our configuration is 100% through environment variables.

**Comment by rix0rrr** [comment]
Can confirm that if you do not have a `~/.aws` directory on your machine and configure using `AWS_DEFAULT_REGION`, `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, `AWS_SESSION_TOKEN`, it fails on my machine as well.

**Comment by scarytom** [comment]
@otaviomacedo no, we don't us an ~/.aws/config file.  Our CI system just sets the AWS credentials via `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` env vars, and our CDK Environment then picks one of our other AWS accounts that has been cdk-bootstrapped to allow access via RBAC.

This setup has been working fine for years, and now fails with 2.167.0

**Comment by otbe** [comment]
> @otbe:
> 
>     1. Do you also have a `~/.aws/config` file?
> 
>     2. Were you running this code on an EC2 instance?

1. no `~/.aws/config` 
2. Yes it was executed on EC2 (however in a docker container)

**Comment by rix0rrr** [comment]
There are 2 distinct problems being discussed in this thread. While they are both introduced by switching from SDKv2 to SDKv3 we should make sure to distinguish them.

- `The security token included in the request is invalid`: seems to happen in some circumstances on CI machines, where configuration happens entirely through environment variables.
- `Unable to determine AWS region from environment or AWS configuration (profile: "default"), defaulting to 'us-east-1'`: the region that is present in a credentials file is not picked up.

This issue was originally created for the first issue by @AdamPD. @otbe, I will make a new issue for your error.

**Comment by rix0rrr** [comment]
On the `The security token included in the request is invalid` error: it occurs to me that the users reporting this seem NOT to be using session credentials, but IAM User credentials instead (access key pair without a session token).

Maybe that's a lead...

**Comment by rix0rrr** [comment]
Woohoo, this reproduces! IAM User credentials seem broken:

```
$ env | grep AWS_
AWS_ACCESS_KEY_ID=*****
AWS_DEFAULT_REGION=us-east-2
AWS_SECRET_ACCESS_KEY=*******

$ npx cdk deploy -vvv
...
[11:33:04] Unable to determine the default AWS account (InvalidClientTokenId): The security token included in the request is invalid
...
Unable to resolve AWS account to use. It must be either configured when you define your CDK Stack, or through the environment
```

**Comment by gsingh1** [comment]
I can confirm the same issue happening on GHE instances too. NPM package deprecation is working as expected and our builds are clearing on `2.166.0`. Thanks.

**Comment by ashishdhingra** [comment]
Another related issues:

**Comment by ashishdhingra** [comment]
Another report https://github.com/aws/aws-cdk/issues/32137

**Comment by mrgrain** [comment]
Hi all, we have released [v2.167.1](https://github.com/aws/aws-cdk/releases/tag/v2.167.1) with fixes for the reported auth issues.

Please try again and report back. Reports that everything is working are equally useful as failures.
